src/cli/build/plugin.js

```diff
@@ -184,7 +196,36 @@ let state = {
     // TODO: When we make the postcss plugin async-capable this can become async
     let files = fastGlob.sync(this.contentPatterns.all)
 
+    // Detect whether a glob pattern might be too broad. This means that it:
+    // - Includes `**`
+    // - Does not include any of the known large directories (e.g.: node_modules)
+    let maybeBroadPattern = this.contentPatterns.all.some(
+      (path) => path.includes('**') && !LARGE_DIRECTORIES_REGEX.test(path)
+    )
+
+    // All globs that explicitly contain any of the known large directories (e.g.:
+    // node_modules)
+    let explicitGlobs = this.contentPatterns.all.filter((path) =>
+      LARGE_DIRECTORIES_REGEX.test(path)
+    )
+
     for (let file of files) {
+      if (
+        maybeBroadPattern &&
+        // When a broad pattern is used, we have to double check that the file was
+        // not explicitly included in the globs.
+        !micromatch.isMatch(file, explicitGlobs)
+      ) {
+        let largeDirectory = LARGE_DIRECTORIES.find((directory) => file.includes(directory))
+        if (largeDirectory) {
+          log.warn('broad-content-glob-pattern', [
+            `You are using a glob pattern that includes \`${largeDirectory}\` without explicitly specifying \`${largeDirectory}\` in the glob.`,
+            'This can lead to performance issues and is not recommended.',
+            'Please consider using a more specific pattern.',
+            'https://tailwindcss.com/docs/content-configuration#pattern-recommendations',
```

Comment:
I think the warning could be a bit clearer by:

- Using terminology from our docs: _Your tailwind settings contain a `content` setting that..._ instead of glob pattern (might not be immediately clear what this means). 
- Giving more hints as to what needs to change. We could include the pattern we found to be faulty: _Your content settings contain a pattern that is too broad: `./**/*.js`. This matches files inside ${largeDirectory} which is ... xyz_

I can imagine this warning popping up for all people in teams that have had their setup for a long time, so it might pop up for people that don't immediately know how the project is set up.

Reply:
Yeah it's mainly a placeholder

- We do talk about glob patterns here though: https://tailwindcss.com/docs/content-configuration#configuring-source-paths
- Including the actual glob pattern might be useful indeed

Reply:
Yeah I think the docs are good but they also explain what glob patterns are and why they are useful. We can be clear and call them "globs inside the `content` setting" or so. Just want to make sure it's very clear what setting we're talking about.